### ISSUE TITLE
Question: Redis 3.0.0 RC already. Does StackExchange.Redis support for Redis 3.0 cluster? if so? how we can config it?

### ISSUE THREAD

**Issue by mrgonext** [StackExchange/StackExchange.Redis#107]
Redis 3.0.0 RC already. Does StackExchange.Redis support for Redis 3.0 cluster? if so? how we can config it?

**Comment by Joe-U-Questionmark** [comment]
@mgravell just to be explicit, does this mean that StackExchange.Redis does fully support Redis 3.0?

Thanks!

**Comment by mgravell** [comment]
In theory: yes. It certainly worked againat the previous builds. However,
due to multiple projects and limited time, I have *not* yet
extensively/explicitly tested against the current 3.0 build, and I know
there are a few tweaks I want to make (that you shouldn't notice - for
example, whether it uses "cluster nodes" vs "cluster slots" to prime the
hash cache)

I'd be happy to hear feedback if you're in a position to try it...

**Comment by robermar2** [comment]
Not sure the current status of this.  Although it does imply that "redis cluster" is supported with the current documentation.  We are going to be trying this out regardless, but can anyone state clearly whether redis cluster with the latest stable build of redis is supported?

**Comment by amanbhushan** [comment]
StackExchange.Redis 1.1.603  version throwing MOVED exception when trying to set or get in redis cluster.

i have initialized ConnectionMultiplexer  with one random picked cluster node

**Comment by rupendra-sharma** [comment]
just wondering if this issue has been fixed? I am facing this issue with v1.2.4 also.

**Comment by mgravell** [comment]
it *should* work simply by connecting to any cluster node; I will need to
investigate against my test cluster.  Any specific cluster size and key you
are seeing this with?

**Comment by rupendra-sharma** [comment]
Hi @mgravell 
It took me some time to test and reproduce the issue. The issue is:

> An exception of type 'StackExchange.Redis.RedisConnectionException' occurred in StackExchange.Redis.dll but was not handled in user code
> 
> Additional information: No connection is available to service this operation: SADD {user}systemuser; An existing connection was forcibly closed by the remote host; IOCP: (Busy=2,Free=998,Min=4,Max=1000), WORKER: (Busy=8,Free=32759,Min=4,Max=32767), Local-CPU: n/a

I am running a redis cluster on windows machine with 6 nodes from 127.0.0.1:7002 to 7007. This issue encountered while I have given  only 1 cluster node 127.0.0.1:7000 (master node) in the Redis connection string. In case any of the other master nodes down, the Redis operation throws socket connection timeout only once but in next try the operation is successful.
But if the master node 127.0.0.1:7000 is down, the Redis code never gets healed and throw the error written above. I am running the StackExchange.Redis and StackExchange.Redis.StrongName ( v 1.2.4) and Microsoft.Web.RedisSessionStateProvider ( v2.2.5).
The connection code is as following:
`private static bool Connect()
        {
            if (_db == null)
            {
                JsonConvert.DefaultSettings = () => new JsonSerializerSettings
                {
                    //ReferenceLoopHandling = ReferenceLoopHandling.Ignore,    // will not serialize an object if it is a child object of itself
                    ReferenceLoopHandling = ReferenceLoopHandling.Serialize,  // is useful if objects are nested but not indefinitely
                    //PreserveReferencesHandling = PreserveReferencesHandling.Objects, // serialize an object that is nested indefinitely
                    TypeNameHandling = TypeNameHandling.All
                };
                var configurationOptions = ConfigurationOptions.Parse(REDIS_URL.ToString());
                configurationOptions.SyncTimeout = 30000;
                if (!string.IsNullOrEmpty(REDIS_PASSWORD))
                    configurationOptions.Password = REDIS_PASSWORD;

                _redis = ConnectionMultiplexer.Connect(configurationOptions);
                _db = _redis.GetDatabase();
            }

            return _db != null;
        }`

And the Redis operation throwing the error is:
` public bool AddToSet(string redisKey, string value)
        {
            return _db.SetAdd(redisKey, value);
        }`

**Comment by NickCraver** [comment]
TL;DR: Bug report is if the first node in the endpoint list is down, things won't work. We need to test this, though adding an automated test is possibly problematic in itself (to accurately test the scenario).

**Comment by NickCraver** [comment]
This was fixed in 2.0 via snapshot changes to solve several issues in the area - closing out to cleanup!

**Comment by toebens** [comment]
hi, from what i have read to far we need to configure the redis cluster with another tool - not stackexchange.redis!? couldn't find any info about it other than in the official docs at https://redis.io/topics/cluster-tutorial can you please confirm?
